### ISSUE TITLE
Join from pandas not matching identical strings 

### ISSUE THREAD

**Issue by bmschmidt** [duckdb/duckdb#1767]
Thank you for this impressive and useful project.

**What does happen?**

When joining two registered pandas dfs including a Unicode character (not sure if it's a Unicode problem, but that's where I've been able to reproduce), the join sometimes fails to match identical strings.

**What should happen?**
The two blocks below should return equivalent results.

**To Reproduce**
```python
import pandas as pd
import duckdb

A = pd.DataFrame({"key": ["a", "п"]})
B = pd.DataFrame({"key": ["a", "п"]})
con = duckdb.connect(":memory:")
arrow = con.register("A", A).register("B", B)
q = arrow.query("""SELECT key FROM "A" FULL JOIN "B" USING ("key")""")
q.df()
```
gives me

```
	key
0	a
1	п
2	п
```

while swapping the order of the values

```python
import pandas as pd
import duckdb
C = pd.DataFrame({"key": ["п", "a"]})
D = pd.DataFrame({"key": ["п", "a"]})
con = duckdb.connect(":memory:")
arrow = con.register("C", C).register("D", D)
q = arrow.query("""SELECT key FROM "C" FULL JOIN "D" USING ("key")""")
q.df()
```

Gives me:

```
key
п
a
```

**Environment (please complete the following information):**
OS X 10.14.3
DuckDB '0.2.7-dev108' (was also getting with the last stable release)
Pandas '1.2.4'

**Before submitting**
- [Yes] Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?
- [Yes] Have you tried this on the latest `master` branch? In case you cannot compile, you may find some binaries here: https://github.com/duckdb/duckdb/releases/tag/master-builds

**Comment by hawkfish** [comment]
(For the visually challenged, that appears to be a pi, not an en.)

**Comment by Mytherin** [comment]
Thanks for the detailed report and investigation! I have pushed a fix for this in #1767. The issue indeed was related to the unicode conversion: we were missing a call to `string_t::Finalize` in the unicode convert loop. This call is required as it adds padding zero's to small inlined strings. When the call is missing, the string will instead contain uninitialized memory there, which can cause the equality comparison to (incorrectly) output that the two instances of the same string are not equal to one another.

I have also expanded the verification of strings to hopefully catch these cases earlier in the future.
